### PR TITLE
Add new mavlink mode EV for external vision

### PR DESCRIPTION
**Pull request by AnnaDaiZH**
**This PR adds an additional mavlink stream mode for external vision.**

Before, Obstacle Avoidance received mavlink messages with onboard mavlink mode. The onboard mode streams far more topics than required by obstacle avoidance. 
Before, VIO was enabled by OSD + required mavlink topics (HIGHRES_IMU, CAMERA_TRIGGER, LOCAL_POSITION_NED) on a custom TX2 board because the bandwidth was limited. Onboard mode caused frequent message dropouts and led to instability. 

Now with the new mavlink mode "External Vision", only the minimal required topics for VIO applications and OA are streamed. This decreases message dropout rate, increases communication reliability and user friendliness for VIO and OA users. 

**To test it**
The mavlink mode can be set in QGC MAV_X_MODE = 8. After reboot, the mavlink streams can be validated in QGC mavlink shell by running 
`mavlink status streams`

@DanielePettenuzzo @dagar 

**Flight logs**
Test flights with the new mavlink stream mode enabling communication to companion computer for obstacle avoidance and collision prevention.
https://review.px4.io/plot_app?log=db0315e8-6268-4806-a896-a838a25af72b
https://review.px4.io/plot_app?log=75c1e58c-85f2-48c4-a865-670a38e151b7

### CODE REVIEW

File: src/modules/mavlink/mavlink_main.cpp [hunk 1]
```diff
@@ -1784,6 +1784,29 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
...
+		configure_stream_local("ATTITUDE_TARGET", 10.0f);
+		configure_stream_local("CAMERA_TRIGGER", unlimited_rate); // for VIO
+		configure_stream_local("CAMERA_IMAGE_CAPTURED", unlimited_rate); //for VIO
+		configure_stream_local("DISTANCE_SENSOR", 10.0f);		// for avoidance
+		configure_stream_local("ESTIMATOR_STATUS", 1.0f);
+		configure_stream_local("EXTENDED_SYS_STATE", 1.0f);
+		configure_stream_local("GLOBAL_POSITION_INT", 10.0f);
+		configure_stream_local("GPS_RAW_INT", 1.0f);
+		configure_stream_local("HOME_POSITION", 0.5f);
+		configure_stream_local("HIGHRES_IMU", 250.0f);		// for VIO
```
Comment (mhkabir) [inline src/modules/mavlink/mavlink_main.cpp]: Set this to unlimited rate. The real rate is around 244 Hz anyway.

File: src/modules/mavlink/mavlink_main.cpp [hunk 2]
```diff
@@ -1784,6 +1784,29 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
 		configure_stream_local("WIND_COV", 10.0f);
 		break;
 
+	case MAVLINK_MODE_EV:
+		configure_stream_local("ALTITUDE", 10.0f); // for avoidance
+		configure_stream_local("ATTITUDE", 25.0f);
+		configure_stream_local("ATTITUDE_TARGET", 10.0f);
+		configure_stream_local("CAMERA_TRIGGER", unlimited_rate); // for VIO
+		configure_stream_local("CAMERA_IMAGE_CAPTURED", unlimited_rate); //for VIO
```
Comment (mhkabir) [inline src/modules/mavlink/mavlink_main.cpp]: This should not be needed.

File: src/modules/mavlink/mavlink_main.cpp [hunk 3]
```diff
@@ -1784,6 +1784,29 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
...
+		configure_stream_local("HOME_POSITION", 0.5f);
+		configure_stream_local("HIGHRES_IMU", 250.0f);		// for VIO
+		configure_stream_local("LOCAL_POSITION_NED", 30.0f); //for VIO and avoidance
+		configure_stream_local("RC_CHANNELS", 5.0f);
+		configure_stream_local("SERVO_OUTPUT_RAW_0", 1.0f);
+		configure_stream_local("SYS_STATUS", 5.0f);
+		configure_stream_local("SYSTEM_TIME", 1.0f);
+		configure_stream_local("TRAJECTORY_REPRESENTATION_WAYPOINTS", 5.0f); // for avoidance
+		configure_stream_local("VFR_HUD", 25.0f);
+		configure_stream_local("WIND_COV", 2.0f);
```
Comment (mhkabir) [inline src/modules/mavlink/mavlink_main.cpp]: 10 Hz TIMESYNC stream is missing.

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: From the tests with rovio, I'd rather keep it at a fixed rate of 250 Hz.

@bkueng @DanielePettenuzzo

File: src/modules/mavlink/mavlink_main.cpp (hunk 2 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: we need those topics to enable the realsense D series cameras

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (mhkabir) [reply src/modules/mavlink/mavlink_main.cpp]: Can you explain the rationale? Mavlink scheduling will never let you hit 250 in reality. On top of that, the actual rate of sensor_combined is lower.

File: src/modules/mavlink/mavlink_main.cpp (hunk 2 above)
Comment (mhkabir) [reply src/modules/mavlink/mavlink_main.cpp]: CAMERA_TRIGGER should be all that is needed. The IMAGE_CAPTURED stream is a much bigger message designed for online geotagging at low rates. (1-5Hz)

Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: okay. i see i misunderstood the topic application

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/mavlink/mavlink_main.cpp]: > Set this to unlimited rate. The real rate is around 244 Hz anyway.

Where are you seeing this?

File: src/modules/mavlink/mavlink_main.h [hunk 4]
```diff
@@ -182,6 +182,9 @@ class Mavlink : public ModuleParams
 		MAVLINK_MODE_IRIDIUM,
 		MAVLINK_MODE_MINIMAL,
 
+		// add External vision
+		MAVLINK_MODE_EV,
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.h]: Why don't you just name it descriptively and drop the comment?

File: src/modules/mavlink/mavlink_main.cpp [hunk 5]
```diff
@@ -1784,6 +1784,29 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
...
+		configure_stream_local("CAMERA_TRIGGER", unlimited_rate); // for VIO
+		configure_stream_local("DISTANCE_SENSOR", 10.0f);		// for avoidance
+		configure_stream_local("ESTIMATOR_STATUS", 1.0f);
+		configure_stream_local("EXTENDED_SYS_STATE", 1.0f);
+		configure_stream_local("GLOBAL_POSITION_INT", 10.0f);
+		configure_stream_local("GPS_RAW_INT", 1.0f);
+		configure_stream_local("HOME_POSITION", 0.5f);
+		configure_stream_local("HIGHRES_IMU", 250.0f);		// for VIO
+		configure_stream_local("LOCAL_POSITION_NED", 30.0f); //for VIO and avoidance
+		configure_stream_local("RC_CHANNELS", 5.0f);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.cpp]: RC_CHANNELS needed?

File: src/modules/mavlink/mavlink_main.cpp [hunk 6]
```diff
@@ -1784,6 +1784,29 @@ Mavlink::configure_streams_to_default(const char *configure_single_stream)
...
+		configure_stream_local("SERVO_OUTPUT_RAW_0", 1.0f);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_main.cpp]: SERVO_OUTPUT_RAW_0 needed?

**Comment by mhkabir** [comment]
We should also use this mode to boost mavlink scheduling prio + run at full rate (reduces latency).

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (mhkabir) [reply src/modules/mavlink/mavlink_main.cpp]: I believe in `uorb top` messages per second.

Comment (dagar) [reply src/modules/mavlink/mavlink_main.cpp]: @mhkabir Open an issue if/when you see that with all the specifics. For the primary IMUs on nearly all of the current boards they should all be a 4 ms integration interval (250 Hz).

File: src/modules/mavlink/mavlink_main.cpp (hunk 6 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: no

File: src/modules/mavlink/mavlink_main.cpp (hunk 5 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: no

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.cpp]: Should the rate set to unlimited_rate or remain 250Hz?

Comment (dagar) [reply src/modules/mavlink/mavlink_main.cpp]: I would leave it unlimited_rate. Weren't going to have to change that assumed 250 Hz for some.

**Comment by dagar** [comment]
> We should also use this mode to boost mavlink scheduling prio + run at full rate (reduces latency).

The biggest difference would be (optionally) polling the `HIGHRES_IMU` topic rather than just sleep.  https://github.com/PX4/Firmware/blob/b23374688bc92586c1bb626e594b9a06ddb7d65b/src/modules/mavlink/mavlink_main.cpp#L2244

File: src/modules/mavlink/module.yaml [hunk 7]
```diff
@@ -39,6 +39,7 @@ parameters:
                 5: Config
                 #6: Iridium # as the user does not need to configure this, hide it from the UI
                 7: Minimal
+                8: ExternalVision
```
Comment (dagar) [inline src/modules/mavlink/module.yaml]: These can have spaces.

**Comment by AnnaDaiZH** [comment]
> > We should also use this mode to boost mavlink scheduling prio + run at full rate (reduces latency).
> 
> The biggest difference would be (optionally) polling the topic rather than just sleep.
> 
> [Firmware/src/modules/mavlink/mavlink_main.cpp](https://github.com/PX4/Firmware/blob/b23374688bc92586c1bb626e594b9a06ddb7d65b/src/modules/mavlink/mavlink_main.cpp#L2244)
> 
> Line 2244 in [b233746](/PX4/Firmware/commit/b23374688bc92586c1bb626e594b9a06ddb7d65b)
> 
>  px4_usleep(_main_loop_delay);

Do I need to implement this or is this a thought on another topic, as I don't understand the mavlink scheduling.

**Comment by mhkabir** [comment]
Let's merge this and fix the scheduling performance in a follow-up.

File: src/modules/mavlink/mavlink_main.cpp [hunk 8]
```diff
@@ -3049,7 +3073,7 @@ Start mavlink on UDP port 14556 and enable the HIGHRES_IMU message with 50Hz:
 	PRINT_MODULE_USAGE_PARAM_STRING('t', "127.0.0.1", nullptr,
 					"Partner IP (broadcasting can be enabled via MAV_BROADCAST param)", true);
 #endif
-	PRINT_MODULE_USAGE_PARAM_STRING('m', "normal", "custom|camera|onboard|osd|magic|config|iridium|minimal",
+	PRINT_MODULE_USAGE_PARAM_STRING('m', "normal", "custom|camera|onboard|osd|magic|config|iridium|minimal|ev",
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp]:
```suggestion
	PRINT_MODULE_USAGE_PARAM_STRING('m', "normal", "custom|camera|onboard|osd|magic|config|iridium|minimal|extvision",
```

File: src/modules/mavlink/mavlink_main.h [hunk 9]
```diff
@@ -181,7 +181,7 @@ class Mavlink : public ModuleParams
 		MAVLINK_MODE_CONFIG,
 		MAVLINK_MODE_IRIDIUM,
 		MAVLINK_MODE_MINIMAL,
-
+		MAVLINK_MODE_EXTVISION,
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.h]: Minor detail: can you leave the empty line before MAVLINK_MODE_COUNT?

**Comment by AnnaDaiZH** [comment]
> Can you please squash the commits into one?

@bkueng squashed and pushed

File: src/modules/mavlink/mavlink_main.h (hunk 9 above)
Comment (AnnaDaiZH) [reply src/modules/mavlink/mavlink_main.h]: The empty line fails the style check from the CI.

**Comment by mhkabir** [comment]
@hamishwillee This will need changes to the companion computer connection docs.

**Comment by mhkabir** [comment]
@vooon This will also need some form of message filtering on mavros side to prevent sending e.g 250Hz IMU data to the GCS by default. Ideas?

**Pull request merged**